input_pill: Add edit input_pill feature.

adnan-td opened the pull request:
Here's a quick overview of what each commit does:

1. Edit input_pills in place by

    - Double clicking on a pill
    - Focusing on a pill (via arrow keys) and then pressing enter

2. Show typeahead for pills in composebox DM recipient box.

Fixes part (3) of #29510 

**Screenshots:**

Edit pill example:

1. By double clicking

![case1](https://github.com/zulip/zulip/assets/78212328/b767ec65-95ca-4ae9-a5a3-8bca7a0333cb)

2. Using the Arrow keys + Enter key

![case2](https://github.com/zulip/zulip/assets/78212328/ccd7d267-f034-48bc-b7b0-89bcee6fb9a6)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

adnan-td commented:
Have completed the requirements mentioned in the issue. Just stuck due to not being able to write proper test for the double click event. Need some guidance for the stub based test.

timabbott commented on web/src/input_pill.ts:
```diff
@@ -438,6 +438,18 @@ export function create<T>(opts: InputPillCreateOptions<T>): InputPillContainer<T
+        store.$parent.on("dblclick", (e) => {
+            e.stopPropagation();
+            const $pill = $(e.target)?.closest(".pill");
+            const text = $pill?.find(".pill-label")?.text()?.trim() ?? "";
```
I'm not sure the `text()` value of the pill will always be what you need to recreate the pill in other inputs, like the compose  box user picker, or the stream subscribers picker that supports users/groups/streams. 

So we might need to rework `InputPillItem` to have a `text_value` field that we populate with the exact original text used to create the pill, maybe as a preparatory refactor commit?

timabbott replied:
This might be a lot easier as a refactor now that `input_pill.ts` is TypeScript.

adnan-td replied:
I have tried in all the pill use cases, this `text` value does the job.

timabbott commented on web/styles/input_pill.css:
```diff
@@ -27,7 +27,6 @@
         &:focus {
             color: hsl(0deg 0% 100%);
             border: 1px solid hsl(176deg 78% 28%);
-            background-color: hsl(176deg 49% 42%);
```
This change should be in its own commit with its own commit message, since it's not directly part of adding the double-click behavior.

timabbott commented:
https://github.com/zulip/zulip/pull/29511/files#r1546651687 may offer a good answer to the testing problem too, since reading data is way simpler.

adnan-td committed:
input_pill: Add edit pill feature.

Adds the double click to edit feature to an input pill
by inserting contenteditable input div.
Provides type ahead for the for the pill being edited.

Fixes part(3) of #29510

adnan-td commented:
Almost complete with the PR, small parts of testing need to be worked upon!

adnan-td committed:
compose_box: Add double click to edit input pill feature.

Double click to edit now works in private dm in compose box.

adnan-td commented:
@sahil839 can you review this?

sahil839 reviewed (commented):
Posted a few comments about naming.

The CSS changes like centering the "x" and all are done in #30121 and that fix seems better, so maybe you can remove those changes. Also, I don't think 4th point in the issue is implemented in this PR, so would be better to use "Fixes part of ..." instead of "Fixes ..." in the commit message.

sahil839 commented:
I also noticed a couple of things when testing -
- When I double click on a pill and then just press enter without changing anything, the text is removed. I think user would expect the pill to be created again. 
 ![pill-edit-enter](https://github.com/zulip/zulip/assets/35494118/2753098c-d595-4594-8917-136742bc3e6a)
- When I double click on a pill and then change it, the new pill is added at the last. I do not have a strong opinion about this but just mentioned it.
![new-pill-added-last](https://github.com/zulip/zulip/assets/35494118/ff5dfc26-24a5-476d-9ba6-e3945813ff70)
- I double click on a pill and remove some characters and then click outside the compose box, the compose box closes. Then I open the compose box by clicking on "New direct message" button and click on the recipient input. The cursor is in the new contenteditable div and not in the one which had the remaining text of previous pill, so pressing backspace does not remove the characters.
![Peek 2024-05-27 21-53](https://github.com/zulip/zulip/assets/35494118/abafddff-ba11-4504-81b2-bcbcdbd082da)

sahil839 commented:
Also, can we just remove the `contenteditable` element once a new pill is created after editing instead of just hiding it.

adnan-td commented on web/src/pill_typeahead.ts:
```diff
@@ -26,22 +25,18 @@ function group_matcher(query: string, item: UserGroupPillData): boolean {
-    },
-): void {
+export function set_up($input: JQuery, pills: CombinedPillContainer, opts: TypeaheadOptions): void {
+    if (!$input) {
```
This condition is used when testing the callback from compose `setupTypeahead` function in
![image](https://github.com/zulip/zulip/assets/78212328/3063d07c-313e-4142-a6d4-b923c9c1dfda)

Removing it causes error (2 times running the set_up method).

adnan-td replied:
When we call `addSetupTypeahead`, it sets an arrow function which is used for setting up typeahead.

![image](https://github.com/zulip/zulip/assets/78212328/21e0fbf7-f573-4831-ad42-2fb506e77a81)

To test this in `pill_typeahead.text.js`, this arrow function is called by passing `null` to prevent the function being called with same parameters twice during testing it.

![image](https://github.com/zulip/zulip/assets/78212328/aafa047e-6324-4399-8648-c9b747e77938)

evykassirer replied:
It's not great to add code here just because tests need it, that can get confusing.

I don't see why you need to call `setup_typeahead(null)`, what happens if you remove it?

adnan-td commented on web/src/input_pill.ts:
```diff
@@ -79,11 +81,13 @@ export type InputPillContainer<T> = {
     getCurrentText: () => string | null;
     is_pending: () => boolean;
     _get_pills_for_testing: () => InputPill<T>[];
+    _get_setup_typeahead: () => SetupTypeahead | undefined;
```
change name to _get_setup_typeahead_for_testing

adnan-td commented on web/src/composebox_typeahead.ts:
```diff
@@ -1281,47 +1281,9 @@ export function initialize_compose_typeahead($element: JQuery<HTMLTextAreaElemen
-        header_html: render_topic_typeahead_hint,
-    });
-
+export function initialize_pm_typeahead($selector: JQuery): void {
```
saw $selector being used in another initialize function. will make a commit to fix that as well

adnan-td commented on web/src/input_pill.ts:
```diff
@@ -482,10 +528,15 @@ export function create<T>(opts: InputPillCreateOptions<T>): InputPillContainer<T
             store.createPillonPaste = callback;
         },
 
+        setupTypeahead(callback) {
```
Changed the name to addSetupTypeahead. Naming this is a bit confusing

adnan-td commented:
> I double click on a pill and remove some characters and then click outside the compose box, the compose box closes. Then I open the compose box by clicking on "New direct message" button and click on the recipient input. The cursor is in the new contenteditable div and not in the one which had the remaining text of previous pill, so pressing backspace does not remove the characters.

Fixed this by making the backspace focus into the `contenteditable` div.

[Screencast from 2024-05-28 20-22-32.webm](https://github.com/zulip/zulip/assets/78212328/283a376c-93c6-47d5-80b7-ea7db2e8c6bd)

adnan-td commented:
Have implemented the pillifying input_field when user clicks out of focus feature by adding a blur event.

sahil839 reviewed (commented):
Posted a few more comments.

I am not sure about the last commit to create the pill when user clicks out of focus. It creates pill even when the name is not complete as the code just simulates pressing enter key and that creates pill considering the option on typeahead is selected.

sahil839 commented:
Also, might be good to have a manual testing review once as would be easy to review the code once we are sure about what behavior is expected.

adnan-td commented on web/src/input_pill.ts:
```diff
@@ -449,6 +471,37 @@ export function create<T>(opts: InputPillCreateOptions<T>): InputPillContainer<T
+                if (keydown_util.is_enter_event(e) && store.setupTypeahead) {
+                    $new_input.hide();
+                    // after the event propagates further, call remove
+                    setTimeout(() => $new_input.remove(), 0);
```
Hide was redundant so fixed it and moved removal logic to `bootstrap_typeahead`.

adnan-td commented on web/src/input_pill.ts:
```diff
@@ -351,6 +363,16 @@ export function create<T>(opts: InputPillCreateOptions<T>): InputPillContainer<T
+                    store.$input.trigger("focus");
+                    return;
+                }
+                if ($(this).prev().attr("editable_pill") === "true") {
```
Now using class name `editable-pill` for simplicity.

adnan-td commented on web/src/input_pill.ts:
```diff
@@ -449,6 +471,37 @@ export function create<T>(opts: InputPillCreateOptions<T>): InputPillContainer<T
+                    store.$input.trigger("focus");
+                }
+                if (e.key === "ArrowLeft" && window.getSelection()?.anchorOffset === 0) {
+                    e.stopPropagation();
```
Removed the event listener on `new_input` and shifted all events to the above handler.

adnan-td commented:
> I am not sure about the last commit to create the pill when user clicks out of focus.
I have removed the commit. I think it should be played around and improved. Will discuss more about requirements and work on it in a separate PR.

adnan-td commented:
Add feature to edit pills by focusing on it with arrow keys and pressing enter as shown:

[Screencast from 2024-06-06 17-57-16.webm](https://github.com/zulip/zulip/assets/78212328/ea43863f-aaa2-4bfe-b35e-90c4dea7743c)

alya commented:
> When I double click on a pill and then change it, the new pill is added at the last. I do not have a strong opinion about this but just mentioned it.

I found this quite disorienting in manual testing. It would be much better if pills could be edited in-place.

alya commented:
Also noted in manual testing:

1. Unpillified text in the DM addressee area should disappear when the compose box is closed, as it does in `main`. Not having that be the case causes very strange results when the compose box is opened again.
![Screenshot 2024-06-06 at 11 09 54](https://github.com/zulip/zulip/assets/2090066/42b3c623-a58a-4f62-91cd-27d197228477)

2. I got into a state with two sets of placeholder text, after deleting the initial text. (I don't remember the exact repro, so I guess you can check the call recording if you don't either and are having a hard time replicating. :))
![Screenshot 2024-06-06 at 11 13 05](https://github.com/zulip/zulip/assets/2090066/e87b0b46-921a-4fd7-8056-a73507b743e6)

adnan-td commented:
> > When I double click on a pill and then change it, the new pill is added at the last. I do not have a strong opinion about this but just mentioned it.
> 
> I found this quite disorienting in manual testing. It would be much better if pills could be edited in-place.

Made changes in the code to allow in place editing of the pills.

Fixed the following bugs:

1. Unpillified text in the DM addressee area should disappear when the compose box is closed, as it does in main. Not having that be the case causes very strange results when the compose box is opened again.
Fixed by modifying the `removeAllPill` function to also remove `editable-pill`.

3. I got into a state with two sets of placeholder text, after deleting the initial text. (I don't remember the exact repro, so I guess you can check the call recording if you don't either and are having a hard time replicating. :))
Fixed by using a separate template for the `editable-pill`.

evykassirer commented:
> Made changes in the code to allow in place editing of the pills.

When I finish editing a pill, I still see it placed at the end of the input instead of where I was editing it.

evykassirer commented:
Here's a bug I found where editing Cordelia places the invalid input as text at the end of of the input field (for some reason I wasn't able to replicate with other users)

![Kapture 2024-06-07 at 11 25 58](https://github.com/zulip/zulip/assets/5634097/ff519551-0047-486c-b550-85134f24dc8f)

evykassirer commented:
If I have invalid input in the stream subscriber modal, and press enter, I see this error. I think instead of showing me the error it should shake the invalid input, the way it would if that input was in the regular text field.

<img width="565" alt="image" src="https://github.com/zulip/zulip/assets/5634097/75d02e1b-05e2-4f82-9e54-b9b64dfa936e">

It does look like the "No user" error shows up on invalid input now too, but the shake animation feels like important for communicating to the user what's going on.

evykassirer commented on web/src/typeahead_helper.ts:
```diff
@@ -39,6 +39,17 @@ export type CombinedPillItem =
+    update_func?: () => void;
+};
+
+export type SetupTypeahead = ($input: JQuery) => void;
```
I think it's better to keep these in `pill_typeahead.ts` where they're being used, and we can move them to a shared file in the future if they're used more widely.

evykassirer commented on web/src/composebox_typeahead.ts:
```diff
@@ -1363,6 +1325,49 @@ export function initialize({
     });
+}
+
+export function initialize({
```
Try keeping `initialize` above `initialize_pm_typeahead`, I think it'll make this diff a lot smaller and easier to read.
